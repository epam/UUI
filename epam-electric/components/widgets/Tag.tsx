import { createSkinComponent } from '@epam/uui-core';
import * as uui from '@epam/uui';

export type TagMods = {
    /**
     * Defines component color.
     * @default 'neutral'
     */
    color?: uui.TagProps['color'] | 'white' | 'night100' | 'night700';
};

/** Represents the properties of a Tag component. */
export type TagProps = uui.TagCoreProps & TagMods;

export const Tag = /* @__PURE__ */createSkinComponent<uui.TagProps, TagProps>(uui.Tag);
