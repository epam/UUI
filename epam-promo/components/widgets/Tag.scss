@use  '../../assets/styles/index' as *;
@import '../../assets/styles/colorvars/widgets/tag-colorvars';

.root {
    &:global(.uui-button-box) {
        background-color: var(--background-color);
        fill: var(--fill);
        border: 0;

        :global(.uui-count) {
            background-color: var(--count-background-color);
            color: var(--count-color);
            font-family: $font-sans-semibold;
            margin: auto;
            height: 18px;
            line-height: 18px;
            font-size: 12px;
            border-radius: 9px;
            padding-left: 6px;
            padding-right: 6px;
        }

        :global(.uui-count):not(:first-child) {
            margin-left: 6px;
        }

        &:global(.-clickable) {
            &:hover {
                background-color: var(--background-color-hover);
                fill: var(--fill-hover);
                cursor: pointer;
            }
        }

        :global(.uui-caption) {
            font-family: $font-sans;
            color: var(--uui-caption);
        }
    }

    &.size-18 {
        min-height: 18px;
        min-width: 18px;
        padding-left: 3px;
        padding-right: 3px;

        :global(.uui-count) {
            height: 12px;
            min-width: 6px;
            line-height: 12px;
            font-size: 8px;
            border-radius: 6px;
            padding-left: 3px;
            padding-right: 3px;
        }

        :global(.uui-count):not(:first-child) {
            margin-left: 3px;
        }

        :global(.uui-caption), :global(.uui-input) {
            padding: 3px 3px;
            line-height: 12px;
            font-size: 12px;
        }
    }

    &.size-24 {
        @include button-horizontal-layout(24px, 0);
        min-height: 24px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 3px;
            padding-bottom: 3px;
            line-height: 18px;
            font-size: 14px;
        }
    }

    &.size-30 {
        @include button-horizontal-layout(30px, 0);
        min-height: 30px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 6px;
            padding-bottom: 6px;
            line-height: 18px;
            font-size: 14px;
        }
    }

    &.size-36 {
        @include button-horizontal-layout(36px, 0);
        min-height: 36px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 6px;
            padding-bottom: 6px;
            line-height: 24px;
            font-size: 14px;
        }
    }

    &.size-42 {
        @include button-horizontal-layout(42px, 0);
        min-height: 42px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 9px;
            padding-bottom: 9px;
            line-height: 24px;
            font-size: 16px;
        }

        :global(.uui-count) {
            height: 24px;
            line-height: 24px;
            font-size: 14px;
            border-radius: 12px;
            min-width: 12px;
        }
    }

    &.size-48 {
        @include button-horizontal-layout(48px, 0);
        min-height: 48px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 12px;
            padding-bottom: 12px;
            line-height: 24px;
            font-size: 16px;
        }

        :global(.uui-count) {
            height: 24px;
            line-height: 24px;
            font-size: 14px;
            border-radius: 12px;
            min-width: 12px;
        }
    }
}
