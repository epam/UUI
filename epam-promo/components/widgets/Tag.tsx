import { withMods } from '@epam/uui-core';
import { Button, ButtonProps } from '@epam/uui-components';
import { systemIcons } from '../../icons/icons';
// import buttonCss from '../buttons/Button.scss';
import styles from '../../assets/styles/colorvars/widgets/tag-colorvars.scss';
import css from './Tag.scss';

const defaultSize = '36';

const mapSize = {
    '48': '48',
    '42': '48',
    '36': '36',
    '30': '30',
    '24': '30',
    '18': '18',
};

export interface TagMods {
    size?: '18' | '24' | '30' | '36' | '42' | '48';
}

export function applyTagMods(mods: TagMods) {
    return [
        // buttonCss.root,
        styles['tag-color'],
        css['size-' + (mods.size || defaultSize)],
        css.root,
    ];
}

export const Tag = withMods<ButtonProps, TagMods>(
    Button,
    applyTagMods,
    (props) => ({
        dropdownIcon: systemIcons[mapSize[props.size] || defaultSize].foldingArrow,
        clearIcon: systemIcons[mapSize[props.size] || defaultSize].clear,
    }),
);
