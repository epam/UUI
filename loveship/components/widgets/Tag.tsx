import { withMods } from '@epam/uui-core';
import * as uui from '@epam/uui';
import * as types from '../../components/types';
import css from './Tag.module.scss';

const defaultSize = '18';

export interface TagMods extends uui.TagMods {
    /**
     * @default 'solid'
     */
    fill?: types.FillStyle;
}

export function applyTagMods(mods: TagMods) {
    return [css.root, css['fill-' + (mods.fill || 'solid')]];
}

export interface TagProps extends Omit<uui.TagProps, 'color'>, TagMods {}

export const Tag = withMods<Omit<TagProps, 'color'>, TagMods>(uui.Tag, applyTagMods, (props) => ({
    size: props.size || defaultSize,
}));
