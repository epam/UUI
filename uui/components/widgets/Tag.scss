@use  '../../assets/styles/index' as *;

.root {
    &:global(.uui-button-box) {
        background-color: var(--uui-tag-bg);
        fill: var(--uui-tag-fill);
        border: 0;

        :global(.uui-count) {
            background-color: var(--uui-tag-count-bg);
            color: var(--uui-tag-count);
            font-family: var(--font-semibold);
            margin: auto;
            height: 18px;
            line-height: 18px;
            font-size: 12px;
            border-radius: 9px;
            padding-left: 6px;
            padding-right: 6px;
        }

        :global(.uui-count):not(:first-child) {
            margin-left: 6px;
        }

        &:global(.-clickable) {
            &:hover {
                background-color: var(--uui-tag-bg-hover);
                fill: var(--uui-tag-fill-hover);
                cursor: pointer;
            }
        }

        :global(.uui-caption) {
            font-family: var(--font-regular);
            color: var(--uui-tag-caption);
        }
    }

    &.size-18 {
        min-height: 18px;
        min-width: 18px;
        padding-left: 3px;
        padding-right: 3px;

        :global(.uui-count) {
            height: 12px;
            min-width: 6px;
            line-height: 12px;
            font-size: 8px;
            border-radius: 6px;
            padding-left: 3px;
            padding-right: 3px;
        }

        :global(.uui-count):not(:first-child) {
            margin-left: 3px;
        }

        :global(.uui-caption), :global(.uui-input) {
            padding: 3px 3px;
            line-height: 12px;
            font-size: 12px;
        }
    }

    &.size-24 {
        @include button-horizontal-layout(24px, 0);
        min-height: 24px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 3px;
            padding-bottom: 3px;
            line-height: 18px;
            font-size: 14px;
        }
    }

    &.size-30 {
        @include button-horizontal-layout(30px, 0);
        min-height: 30px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 6px;
            padding-bottom: 6px;
            line-height: 18px;
            font-size: 14px;
        }
    }

    &.size-36 {
        @include button-horizontal-layout(36px, 0);
        min-height: 36px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 6px;
            padding-bottom: 6px;
            line-height: 24px;
            font-size: 14px;
        }
    }

    &.size-42 {
        @include button-horizontal-layout(42px, 0);
        min-height: 42px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 9px;
            padding-bottom: 9px;
            line-height: 24px;
            font-size: 16px;
        }

        :global(.uui-count) {
            height: 24px;
            line-height: 24px;
            font-size: 14px;
            border-radius: 12px;
            min-width: 12px;
        }
    }

    &.size-48 {
        @include button-horizontal-layout(48px, 0);
        min-height: 48px;

        :global(.uui-caption), :global(.uui-input) {
            padding-top: 12px;
            padding-bottom: 12px;
            line-height: 24px;
            font-size: 16px;
        }

        :global(.uui-count) {
            height: 24px;
            line-height: 24px;
            font-size: 14px;
            border-radius: 12px;
            min-width: 12px;
        }
    }
}
