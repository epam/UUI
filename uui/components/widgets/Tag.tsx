import React from 'react';
import { CX, cx, devLogger, Icon, IDropdownToggler, IHasCaption, IHasIcon, uuiElement, uuiMarkers } from '@epam/uui-core';
import { Clickable, ClickableComponentProps, IconContainer } from '@epam/uui-components';
import { CountIndicator, CountIndicatorProps } from './CountIndicator';
import { systemIcons } from '../../icons/icons';
import css from './Tag.module.scss';

const DEFAULT_SIZE = '36';

const mapCountIndicatorSizes: Record<TagCoreProps['size'], CountIndicatorProps['size']> = {
    18: '12',
    24: '18',
    30: '18',
    36: '18',
    42: '24',
    48: '24',
};

interface TagMods {
    /**
     * Defines component color.
     * @default 'neutral'
     */
    color?: 'info' | 'success' | 'warning' | 'critical' | 'neutral';
}

/** Represents the Core properties of the Tag component. */
export type TagCoreProps = ClickableComponentProps & IDropdownToggler & IHasIcon & IHasCaption & {
    /**
     * Defines component size.
     * @default '36'
     */
    size?: '18' | '24' | '30' | '36' | '42' | '48';
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: 'solid' | 'outline';
    /** Call to clear toggler value */
    onClear?(e?: any): void;
    /** Icon for clear value button (usually cross) */
    clearIcon?: Icon;
    /**
     * CSS classes to put on the caption
     * @deprecated
     * */
    captionCX?: CX;
    /** Icon for drop-down toggler */
    dropdownIcon?: Icon;
    /** Count value to be placed in component */
    count?: React.ReactNode;
};

/** Represents the properties of the Tag component. */
export type TagProps = TagCoreProps & TagMods;

function applyTagMods(props: TagProps) {
    return [
        css['size-' + (props.size || DEFAULT_SIZE)],
        css.root,
        `uui-color-${props.color || 'neutral'}`,
        `uui-fill-${props.fill || 'solid'}`,
        'uui-tag',
    ];
}

export const Tag = /* @__PURE__ */React.forwardRef<HTMLButtonElement | HTMLAnchorElement | HTMLSpanElement, TagProps>((props, ref) => {
    if (__DEV__ && props.captionCX) {
        devLogger.warn('Tag: Property \'captionCX\' is deprecated and will be removed in the future release. Please use \'cx\' prop to access caption styles and use cascading to change the styles for the \'uui-caption\' global class');
    }

    const styles = [applyTagMods(props), props.cx];

    const ClearIcon = props.clearIcon ? props.clearIcon : systemIcons.clear;
    const DropdownIcon = props.dropdownIcon ? props.dropdownIcon : systemIcons.foldingArrow;

    return (
        <Clickable
            { ...props }
            rawProps={ {
                'aria-haspopup': props.isDropdown,
                'aria-expanded': props.isOpen,
                ...props.rawProps,
            } }
            cx={ styles }
            ref={ ref }
        >
            { props.icon && props.iconPosition !== 'right' && (
                <IconContainer
                    icon={ props.icon }
                    onClick={ !props.isDisabled ? props.onIconClick : undefined }
                />
            ) }
            { props.caption && (
                <div className={ cx(uuiElement.caption, props.captionCX) }>
                    { props.caption }
                </div>
            ) }
            { props.count !== undefined && props.count !== null && (
                <CountIndicator
                    color={ (!props.color || props.color === 'neutral') ? 'white' : props.color }
                    size={ mapCountIndicatorSizes[props.size || DEFAULT_SIZE] }
                    caption={ props.count }
                />
            ) }
            { props.icon && props.iconPosition === 'right' && (
                <IconContainer icon={ props.icon } onClick={ !props.isDisabled ? props.onIconClick : undefined } />
            ) }
            { props.isDropdown && (
                <IconContainer icon={ DropdownIcon } flipY={ props.isOpen } />
            )}
            { props.onClear && !props.isDisabled && (
                <IconContainer cx={ uuiMarkers.clickable } icon={ ClearIcon } onClick={ props.onClear } />
            ) }
        </Clickable>
    );
});
