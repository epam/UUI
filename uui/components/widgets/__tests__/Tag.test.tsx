import React from 'react';
import { Tag } from '../Tag';
import { renderSnapshotWithContextAsync, renderWithContextAsync, screen, userEvent } from '@epam/uui-test-utils';
// @ts-expect-error TODO: fix the error about unresolved import.
// eslint-disable-next-line import/no-unresolved
import { ReactComponent as CalendarIcon } from '@epam-assets/icons/action-calendar-fill.svg';

describe('Tag', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(<Tag />);
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <Tag
                caption="Test badge"
                icon={ CalendarIcon }
                count={ 12 }
                onIconClick={ () => {} }
                onClick={ () => {} }
                onClear={ () => {} }
                size="36"
            />,
        );
        expect(tree).toMatchSnapshot();
    });

    it('should call `onClear` callback when activated using keyboard navigation', async () => {
        const onClearMock = jest.fn();

        await renderWithContextAsync(
            <Tag
                caption="Test badge"
                icon={ CalendarIcon }
                onClear={ onClearMock }
            />,
        );

        const tag = await screen.findByText(/test badge/i);
        expect(tag).toBeInTheDocument();

        const clearButton = await screen.findByRole(
            'button',
            {
                name: /remove tag/i,
            },
        );
        expect(clearButton).toBeInTheDocument();
        expect(clearButton).not.toHaveFocus();

        await userEvent.click(tag);
        await userEvent.tab();
        expect(clearButton).toHaveFocus();

        await userEvent.keyboard('{Enter}');
        expect(onClearMock).toBeCalled();
    });

    it('calls onIconClick', async () => {
        const onIconClick = jest.fn();
        await renderWithContextAsync(
            <Tag
                caption="Test badge"
                icon={ CalendarIcon }
                onIconClick={ onIconClick }
            />,
        );
        const clickableIcon = screen.getByRole(
            'button',
            {
                name: /icon in input/i,
            },
        );

        await userEvent.click(clickableIcon);

        expect(onIconClick).toBeCalledTimes(1);
    });
});
