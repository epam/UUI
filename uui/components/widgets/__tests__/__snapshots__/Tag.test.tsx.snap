// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag should be rendered correctly 1`] = `
<DocumentFragment>
  <span
    class="uui-button-box uui-enabled root uui-size-36 uui-color-neutral uui-fill-solid uui-tag"
    tabindex="-1"
  />
</DocumentFragment>
`;

exports[`Tag should be rendered correctly with props 1`] = `
<DocumentFragment>
  <button
    class="uui-button-box uui-enabled -clickable root uui-size-36 uui-color-neutral uui-fill-solid uui-tag"
    tabindex="0"
    type="button"
  />
  <button
    aria-label="Icon in input"
    class="uui-button-box uui-enabled -clickable container uui-icon_button uui-color-neutral root"
    tabindex="0"
    type="button"
  >
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
  </button>
  <div
    class="uui-caption"
  >
    Test badge
  </div>
  <div
    class="root uui-count_indicator uui-size-18 uui-color-white"
  >
    12
  </div>
  <button
    aria-label="Remove tag"
    class="uui-button-box uui-enabled -clickable container uui-icon_button uui-color-neutral root uui-icon-cancel"
    tabindex="0"
    type="button"
  >
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
  </button>
</DocumentFragment>
`;
